rust/feature-flags/src/cohort_operations.rs

```diff
@@ -0,0 +1,369 @@
+use std::collections::HashSet;
+use std::sync::Arc;
+use tracing::instrument;
+
+use crate::cohort_models::{Cohort, CohortId, CohortProperty, InnerCohortProperty};
+use crate::{api::FlagError, database::Client as DatabaseClient, flag_definitions::PropertyFilter};
+
+impl Cohort {
+    /// Returns a cohort from postgres given a cohort_id and team_id
+    #[instrument(skip_all)]
+    pub async fn from_pg(
+        client: Arc<dyn DatabaseClient + Send + Sync>,
+        cohort_id: i32,
+        team_id: i32,
+    ) -> Result<Cohort, FlagError> {
+        let mut conn = client.get_connection().await.map_err(|e| {
+            tracing::error!("Failed to get database connection: {}", e);
+            // TODO should I model my errors more generally?  Like, yes, everything behind this API is technically a FlagError,
+            // but I'm not sure if accessing Cohort definitions should be a FlagError (vs idk, a CohortError?  A more general API error?)
+            FlagError::DatabaseUnavailable
+        })?;
+
+        let query = "SELECT id, name, description, team_id, deleted, filters, query, version, pending_version, count, is_calculating, is_static, errors_calculating, groups, created_by_id FROM posthog_cohort WHERE id = $1 AND team_id = $2";
+        let cohort = sqlx::query_as::<_, Cohort>(query)
+            .bind(cohort_id)
+            .bind(team_id)
+            .fetch_optional(&mut *conn)
+            .await
+            .map_err(|e| {
+                tracing::error!("Failed to fetch cohort from database: {}", e);
+                FlagError::Internal(format!("Database query error: {}", e))
+            })?;
+
+        cohort.ok_or_else(|| {
+            FlagError::CohortNotFound(format!(
+                "Cohort with id {} not found for team {}",
+                cohort_id, team_id
+            ))
+        })
+    }
+
+    /// Returns all cohorts for a given team
+    #[instrument(skip_all)]
+    pub async fn list_from_pg(
+        client: Arc<dyn DatabaseClient + Send + Sync>,
+        team_id: i32,
+    ) -> Result<Vec<Cohort>, FlagError> {
+        let mut conn = client.get_connection().await.map_err(|e| {
+            tracing::error!("Failed to get database connection: {}", e);
+            FlagError::DatabaseUnavailable
+        })?;
+
+        let query = "SELECT id, name, description, team_id, deleted, filters, query, version, pending_version, count, is_calculating, is_static, errors_calculating, groups, created_by_id FROM posthog_cohort WHERE team_id = $1";
+        let cohorts = sqlx::query_as::<_, Cohort>(query)
```

Comment:
An aside but I'm happy to log on a bit late if you want to sync up for an hour some day next week, I can show you how to get sqlx query macros working nicely with CI / your local dev flow - we're finding them super useful over in error tracking land because they give errors at compile time (and therefor also from rust analyser while writing the code) if there's a problem with a query or a struct definition, letting you skip writing a lot of tests asserting simple queries are correct. I wouldn't go re-writing all the existing ones in this codebase, but for new ones you might find it handy. Feel free to throw something in my calendar.

Reply:
I'd love that; I'm in Austin next week so I'll be even closer to you timezone-wise.  Grabbed some time on Wednesday